### ISSUE TITLE
Increase JPEG quality on internal generated thumbnails

### ISSUE THREAD

**Issue by rfgamaral** [Sonarr/Sonarr#616]
Sonarr has a really nice interface but I find that the compression on the generated series posters/fanart/banners is way too way; making the thumbnails look really ugly (pixelated and with noticeable JPEG artifacts).

Space for these is probably not an issue in the machines where Sonarr ends up being installed, so I'd recommend to just turn off compression altogether or just set it to a really high value, like 90 (assuming 100 = highest possible quality, lowest compression).

**Comment by markus101** [comment]
Could you provide some examples?

I'm not seeing any artifacting on posters and the fanart on series details is served as-is.

Which OS?
Which version on mono (if applicable)?

**Comment by Taloth** [comment]
Probably mono coz there we are limited to the encoder in libgdiplus, which we know isn't that great.
Later we plan to generate HQ thumbs server-side. (We can't use stuff like imagemagick coz Sonarr runs on many platforms)

**Comment by rfgamaral** [comment]
I'm on Debian Wheezy, with:

```
Mono JIT compiler version 4.0.1 (tarball Thu May 28 09:08:28 UTC 2015)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        LLVM:          supported, not enabled.
        GC:            sgen
```

Here's a PNG screenshot of my series posters view:

![sonarr_posters](https://cloud.githubusercontent.com/assets/96476/8148883/55580d30-12a8-11e5-832c-653ca194844e.png)

And here's one of the generated posters (original file):

![poster-250](https://cloud.githubusercontent.com/assets/96476/8148890/96fc2c1c-12a8-11e5-9eee-27d1a14d0654.jpg)

So, you're saying this doesn't happen on Windows? Hopefully is some sort of setting that can be tweaked (even if it's on the source code) and not the library that just can't do better.

**Comment by Taloth** [comment]
Library limitation, nothing we can do atm until we do the thumbs server side.

**Comment by rfgamaral** [comment]
What do you mean server side? What exactly are you doing right now?

**Comment by Taloth** [comment]
Atm we use gdiplus, which on windows has full hq bicubic filtering but on mono (libgdiplus) it's limited to bilinear. As well as some other limitations, such as no jpeg quality setting.

All series/episode information is retrieved from a proxy (skyhook) which takes data from thetvdb and enhances it a bit. and subsequently stored in the Sonarr database.
The posters, however, are retrieved straight from thetvdb at full size. The thumbs are generated by Sonarr (with all limitations mentioned earlier).
Ultimately (at some unspecified future date) we would want to do the images in skyhook as well, so we can generate hq thumbs, as well as support for webp.

As for the artifacts, even after we get it implemented in skyhook, there's still gotta be a trade-off for image quality vs size.
The main purpose of the thumbs is to reduce the bandwidth usage, especially on mobile.

Examples:
Current: Mono vs Windows@90: (10kb vs 15kb)
![image](https://cloud.githubusercontent.com/assets/6383890/8150181/5a72756c-12df-11e5-8344-93ea09f1b4f2.png)![poster-250at90](https://cloud.githubusercontent.com/assets/6383890/8150179/4956ad7a-12df-11e5-97b8-9da70b4254a5.jpg)

On windows I can increase Quality to 98, for example (27kb)
![poster-250at98](https://cloud.githubusercontent.com/assets/6383890/8150182/73c85522-12df-11e5-9572-abf0e8b1da6d.jpg)
But this doubles the filesize. Which means that at 200 shows that's going to be 3MB extra. (Still nothing compared to the fullsize images, 30mb, we were serving earlier, but 3MB still hurts on mobile)

Anyway, nothing we can do about the mono quality atm. BUT if you install imagemagick and replace the poster-250.jpg files in the ~/.config/NzbDrone/MediaCover/... folders you can work around it.

For the record, at nominal viewing distance most ppl won't notice the difference.

**Comment by rfgamaral** [comment]
Thanks for the clarification :)

Just out of curiosity, if you don't mind having this discussion...

I looked through the source code before posting this here and though you were using a library called ImageResizer. Is this not true or is this the one using gdiplus to do all the heavy lifting?

However, what's the reason for using skyhook to generate thumbnails on a remote server instead of the local one? Is it because there's no library for image manipulation compatible with all platforms that Sonarr supports? Or something like that? Isn't there an alternative to ImageMagick that could run all platforms and do all the necessary work locally?

> As for the artifacts, even after we get it implemented in skyhook, there's still gotta be a trade-off for image quality vs size. The main purpose of the thumbs is to reduce the bandwidth usage, especially on mobile.

I understand that some people may want to connect to their Sonarr remotely and this can be a concern to them, but personally, I rarely use mobile for these things. So, might I recommend that you add an advanced setting for the quality compression whenever you guys decide to implement this feature?

> BUT if you install imagemagick and replace the poster-250.jpg files in the ~/.config/NzbDrone/MediaCover/... folders you can work around it.

Won't they get overwritten if I press the "update library" button or "update series info"? Or those actions will only add missing info/thumbnails and won't overwrite anything? If not, I might look into creating some sort of script for this :)

> For the record, at nominal viewing distance most ppl won't notice the difference.

Yeah, just like most of my friends don't even care for HD content and watch just about anything. There's no hope for them. But, I'm not most people and for better or worse, I have a really good eye and notice these kind of things. And they really bother me to be honest (I like pixel perfect things). I noticed the bad quality as soon as I added my first series and that was when I decided to post this issue.

But I understand how low priority this is, I can relate :)

**Comment by Taloth** [comment]
Yes, ImageResizer uses gdiplus internally. Imagine my disappointment after adding the lib to Sonarr only to discover that it uses gdiplus under the hood.

Not all platforms which sonarr runs on have ready-made packages for stuff like imagemagick, even libgdiplus isn't available on all of em. And there is no free managed library that does it, we searched, I tried ImageResizer but you know how that went. :cry: Still, doing it in skyhook opens up a few options.

From our perspective any extra config option is not preferable. If we do it on skyhook we'll have to decide on an acceptable compromise. I would probably go for the 95 or 98 option.

The resized images will only be created if they don't exist OR if the full size image was changed. So, yes, whip up a script, and put it in cron. You can probably use ctime/mtime and ```touch``` to check which ones you generated.
It isn't perfect, but it's the best option for you atm.

I'm going to close this Issue since getting it in skyhook is already on the board somewhere.

**Comment by florisrobbemont** [comment]
Just wanted to offer some info.

ImageResizer has a new replacement for gdiplus completely written in C. I dont't know how it'll work on Mono, but it might be worth a shot (not much rewritting involved). It's for V4 of their product and it's almost done (beta now).

Verstuurd vanaf mijn iPhone

**Comment by Taloth** [comment]
Thank you, but sadly that won't work. Sonarr gets installed on exotic platforms like raspberry pi, a variety of synology nas devices, and the list goes on. That's why we want to move it server-side.

**Comment by florisrobbemont** [comment]
That's a good reason :) didn't think of that.

You can still use ImageResizer on the server though :p and perhaps a CDN to reduce bandwidth. But I'm guessing you already thought of that :)

Verstuurd vanaf mijn iPhone

**Comment by rfgamaral** [comment]
Thanks @Talot. The 95 or 98 would work fine for me. If I had to guess, I'd say the compression I see right now would be about 60 or something, so 90+ would be a huge improvement.

I'm no expert on bash scripting, but I'll see what I can do. It will be a fun exercise :)

**Comment by rfgamaral** [comment]
For the record, here's the command I ended up with (I had help from #bash@freenode):

`find /usr/share/sonarr/MediaCover/ -type f -mtime -1 -name '*-*.jpg' -exec bash -c 'for path do dir=${path%/*} file=${path##*/} height=${file%.*}; convert "$dir/${file%-*}.jpg" -resize "x${height#*-}" -unsharp 0x0.5+0.5+0.008 -quality 95 "$path"; touch -r "$dir/${file%-*}.jpg" "$path"; done' _ {} +`

I runs daily with crontab. However, the script assumes that it runs every single day and that can be a problem (not likely). For instance, say I add a new series and metadata is generated. Then I had a power failure and the computer boots only after +24 hours, the script above won't convert those files, because they are older than 1 day and the script skipped one execution.

I'm not sure exactly how to work around this problem. Setting `mtime` to a bigger number will convert the files repeatedly/daily when they were already converted.

Anyway, open to suggestions for improvements :)
